Changes in module 'nf-core/cellranger/count'
--- modules/nf-core/cellranger/count/meta.yml
+++ modules/nf-core/cellranger/count/meta.yml
@@ -30,7 +30,7 @@
   - outs:
       type: file
       description: Files containing the outputs of Cell Ranger
-      pattern: "sample-${meta.gem}/outs/*"
+      pattern: "${meta.gem}/outs/*"
   - versions:
       type: file
       description: File containing software version

--- modules/nf-core/cellranger/count/main.nf
+++ modules/nf-core/cellranger/count/main.nf
@@ -14,8 +14,8 @@
     path  reference
 
     output:
-    tuple val(meta), path("sample-${meta.gem}/outs/*"), emit: outs
-    path "versions.yml"                               , emit: versions
+    tuple val(meta), path("${meta.gem}/outs/*"),        emit: outs
+    path "versions.yml",				emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -27,12 +27,13 @@
     """
     cellranger \\
         count \\
-        --id='sample-${meta.gem}' \\
+        --id='${meta.gem}' \\
         --fastqs=. \\
         --transcriptome=$reference_name \\
         --sample=$sample_arg \\
         --localcores=$task.cpus \\
         --localmem=${task.memory.toGiga()} \\
+        --r1-length=26 \\
         $args
 
     cat <<-END_VERSIONS > versions.yml
@@ -43,8 +44,8 @@
 
     stub:
     """
-    mkdir -p "sample-${meta.gem}/outs/"
-    touch sample-${meta.gem}/outs/fake_file.txt
+    mkdir -p "${meta.gem}/outs/"
+    touch ${meta.gem}/outs/fake_file.txt
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
